lib/net/x86-64-linux/Net.v3

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 Kunal Sareen. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+// Implementation of the "Net" component for Linux, using underlying Linux
+// syscalls and methods.
+
+// XXX: Unsound as we need to enforce that ipv4 and ipv6 are exactly 4 and 16
+// u8s long respectively. One way to achieve this could be to create two more
+// types, Inet4Array and Inet6Array, like so:
+// class Inet4Array {
+// 	def addr: Array<u8> = Array.new(4);
+// }
+// class Inet6Array {
+// 	def addr: Array<u8> = Array.new(16);
+// }
+// type InetAddress {
+// 	case Inet4(addr: Inet4Array);
+// 	case Inet6(addr: Inet6Array);
+// }
+/**
+ * Represents an internet address. It can be one of Inet4 for IPv4 addresses or
+ * Inet6 for IPv6 addresses.
+ */
+type InetAddress {
+	case Inet4(addr: Array<u8>);
+	case Inet6(addr: Array<u8>);
+
+	def render(buf: StringBuilder) -> StringBuilder {
+		match (this) {
+			Inet4(addrV4) => {
+				buf.put3("%d.%d.%d.", addrV4[0], addrV4[1], addrV4[2]);
+				buf.put1("%d", addrV4[3]);
+			}
+			Inet6(addrV6) => {
+				def len = addrV6.length;
+				var zeroStart = -1, zeroEnd = -1;
+				// Find the start and end of the longest chain of zeroes
+				for (i = 0; i < len; i += 2) {
+					var j = i;
+					while (j < len) {
+						if (addrV6[j] == 0 && addrV6[j + 1] == 0) {
+							j += 2;
+						} else {
+							break;
+						}
+					}
+
+					if (j > i && (j - i) > zeroEnd - zeroStart) {
+						zeroStart = i;
+						zeroEnd = j;
+						i = j;
+					}
+				}
+
+				// If only atmost 2 bytes are 0, then we just display them
+				if (zeroEnd - zeroStart <= 2) {
+					zeroStart = -1;
+					zeroEnd = -1;
+				}
+
+				for (i = 0; i < len; i += 2) {
+					if (i == zeroStart) {
+						buf.puts("::");
+						i = zeroEnd;
+						if (i >= len) {
+							break;
+						}
+					} else if (i > 0) {
+						buf.puts(":");
+					}
+					buf.put1("%x", (int.view(addrV6[i]) << 8) | addrV6[i + 1]);
+				}
+			}
+		}
+		return buf;
+	}
+}
+
+class IntPointer(x: int) { }
+
+def socketToSockAddr(socket: SocketBase, server: bool) -> Array<byte> {
+	match (socket.addr) {
+		Inet4(addrV4) => {
+			// Definition of sockaddr for IPv4:
+			// struct sockaddr_in {
+			//   short int          sin_family;  // Address family, AF_INET
+			//   unsigned short int sin_port;    // Port number
+			//   struct in_addr     sin_addr;    // Internet address
+			//   unsigned char      sin_zero[8]; // Same size as struct sockaddr
+			// };
+			def sockAddrBldr = DataWriter.new();
+			def port = if(server, socket.localPort, socket.port);
+			sockAddrBldr.acquire(LinuxConst.SOCKADDR_V4_SIZE);
+			def addrInt = (int.view(addrV4[0]) << 24)
```

Comment:
Hmm, if the address is already a byte array, maybe we can get away with just copying the bytes of it over, instead of packing it into an `int` and then calling `put_b32be`?

Reply:
Done.